File: src/main/java/hdzi/editstarters/buildsystem/gradle/GradleSyntax.java
package hdzi.editstarters.buildsystem.gradle;

import com.intellij.psi.PsiElement;
import hdzi.editstarters.buildsystem.ProjectFile;
import hdzi.editstarters.dependency.DependencyScope;
import hdzi.editstarters.initializr.InitializrBom;
import hdzi.editstarters.initializr.InitializrRepository;
import hdzi.editstarters.initializr.StarterInfo;
import lombok.Data;
import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.tuple.Pair;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public abstract class GradleSyntax<T extends PsiElement> extends ProjectFile<T> {
    @Data
    public static class Instruction {
        private final String inst;
        private final String point;

        public Instruction(String inst, String point) {
            this.inst = inst;
            this.point = point;
        }

        public String toInstString(String format) {
            return format.replace("$inst", inst).replace("$point", point);
        }
    }

    public static final String TAG_DEPENDENCY_MANAGEMENT = "dependencies";
    public static final String TAG_BOM_MANAGEMENT = "dependencyManagement";
    public static final String TAG_BOM_IMPORT = "imports";
    public static final String TAG_BOM = "mavenBom";
    public static final String TAG_REPOSITORY_MANAGEMENT = "repositories";
    public static final String TAG_REPOSITORY = "maven";

    // scope相关指令
    public static final String INS_ANNOTATION_PROCESSOR = "annotationProcessor";
    public static final String INS_COMPILE = "implementation";
    public static final String INS_COMPILE_ONLY = "compileOnly";
    public static final String INS_PROVIDED = "providedRuntime";
    public static final String INS_RUNTIME = "runtimeOnly";
    public static final String INS_TEST = "testImplementation";


    protected List<Instruction> dependencyInstruction(StarterInfo info) {
        String instruction = resolveScope(info.getScope());
        String point = info.getGroupId() + ":" + info.getArtifactId()
                + (info.getVersion() != null ? ":" + info.getVersion() : "");
        List<Instruction> instructions = new ArrayList<>();
        instructions.add(new Instruction(instruction, point));
        // 额外指令
        if (info.getScope() == DependencyScope.ANNOTATION_PROCESSOR) {
            instructions.add(new Instruction(INS_COMPILE_ONLY, point));
        }
        return instructions;
    }

    protected Instruction bomInstruction(InitializrBom bom) {
        return new Instruction(TAG_BOM, String.join(":", bom.getGroupId(), bom.getArtifactId(), bom.getVersion()));
    }

    protected Instruction repositoryInstruction(InitializrRepository repository) {
        return new Instruction(TAG_REPOSITORY, repository.getUrl());
    }

    protected Pair<String, String> splitGroupArtifact(String point) {
        if (StringUtils.isNoneBlank(point)) {
            Matcher matcher = Pattern.compile("^([^:]+):([^:]+)").matcher(point);
            if (matcher.find()) {
                return Pair.of(matcher.group(1), matcher.group(2));
            }
        }
        return Pair.of("", "");
    }

    protected String resolveScope(DependencyScope scope) {
        switch (scope) {
            case ANNOTATION_PROCESSOR:
                return INS_ANNOTATION_PROCESSOR;
            case COMPILE:
                return INS_COMPILE;
            case COMPILE_ONLY:
                return INS_COMPILE_ONLY;
            case PROVIDED:
                return INS_PROVIDED;
            case RUNTIME:
                return INS_RUNTIME;
            case TEST:
                return INS_TEST;
            default:
                return null;
        }
    }

    protected String trimText(String s, char... chars) {
        if (s == null) {
            return s;
        }

        int len = s.length();
        int st = 0;
        char[] val = s.toCharArray();

        while ((st < len) && (ArrayUtils.contains(chars, val[st]))) {
            st++;
        }
        while ((st < len) && (ArrayUtils.contains(chars, val[st]))) {
            len--;
        }
        return ((st > 0) || (len < s.length())) ? s.substring(st, len) : s;
    }
}

File: src/main/kotlin/hdzi/editstarters/buildsystem/gradle/GradleSyntax.kt
package hdzi.editstarters.buildsystem.gradle

import com.intellij.psi.PsiElement
import hdzi.editstarters.buildsystem.ProjectFile
import hdzi.editstarters.springboot.initializr.DependencyScope
import hdzi.editstarters.springboot.initializr.InitializrBom
import hdzi.editstarters.springboot.initializr.InitializrRepository
import hdzi.editstarters.springboot.initializr.StarterInfo

abstract class GradleSyntax<T : PsiElement> : ProjectFile<T>() {
    class Instruction(val instruction: String, val point: String) {
        operator fun component1() = instruction
        operator fun component2() = point
    }

    companion object {
        // 依赖相关指令tag
        const val TAG_DEPENDENCY_MANAGEMENT = "dependencies"
        const val TAG_BOM_MANAGEMENT = "dependencyManagement"
        const val TAG_BOM_IMPORT = "imports"
        const val TAG_BOM = "mavenBom"
        const val TAG_REPOSITORY_MANAGEMENT = "repositories"
        const val TAG_REPOSITORY = "maven"

        // scope相关指令
        const val INS_ANNOTATION_PROCESSOR = "annotationProcessor"
        const val INS_COMPILE = "implementation"
        const val INS_COMPILE_ONLY = "compileOnly"
        const val INS_PROVIDED = "providedRuntime"
        const val INS_RUNTIME = "runtimeOnly"
        const val INS_TEST = "testImplementation"
    }

    protected fun dependencyInstruction(info: StarterInfo): List<Instruction> {
        val instruction = info.scope.resolveScope()
        val version = if (info.version != null) ":${info.version}" else ""
        val point = "${info.groupId}:${info.artifactId}$version"
        val instructions = mutableListOf(Instruction(instruction, point))
        // 额外指令
        if (info.scope == DependencyScope.ANNOTATION_PROCESSOR) {
            instructions.add(Instruction(INS_COMPILE_ONLY, point))
        }
        return instructions
    }

    protected fun bomInstruction(bom: InitializrBom): Instruction {
        val point = "${bom.groupId}:${bom.artifactId}:${bom.version}"
        return Instruction(TAG_BOM, point)
    }

    protected fun repositoryInstruction(repository: InitializrRepository) =
        Instruction(TAG_REPOSITORY, repository.url)

    protected fun splitGroupArtifact(point: String?): Pair<String, String> {
        val group = "^([^:]+):([^:]+)".toRegex().find(point ?: "")?.groupValues
        return Pair(group?.get(1) ?: "", group?.get(2) ?: "")
    }

    private fun DependencyScope.resolveScope() = when (this) {
        DependencyScope.ANNOTATION_PROCESSOR -> INS_ANNOTATION_PROCESSOR
        DependencyScope.COMPILE -> INS_COMPILE
        DependencyScope.COMPILE_ONLY -> INS_COMPILE_ONLY
        DependencyScope.PROVIDED -> INS_PROVIDED
        DependencyScope.RUNTIME -> INS_RUNTIME
        DependencyScope.TEST -> INS_TEST
    }
}